compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -536,9 +547,162 @@ private STToken processNumericLiteral(int startChar) {
 
     /**
      * <p>
-     * Process and returns a hex integer literal.
+     * Process and returns a decimal floating point literal.
+     * </p>
+     * <code>
+     * DecimalFloatingPointNumber :=
+     *    DecimalNumber Exponent [FloatingPointTypeSuffix]
+     *    | DottedDecimalNumber [Exponent] [FloatingPointTypeSuffix]
+     *    | DecimalNumber FloatingPointTypeSuffix
+     * <br/>
+     * DottedDecimalNumber := DecimalNumber . Digit* | . Digit+
+     * <br/>
+     * FloatingPointTypeSuffix := DecimalTypeSuffix | FloatTypeSuffix
+     * <br/>
+     * DecimalTypeSuffix := d | D
+     * <br/>
+     * FloatTypeSuffix :=  f | F
+     * </code>
+     *
+     * @param isDotStart Whether process starts from a dot
+     * @return The decimal floating point literal.
+     */
+    private STToken processDecimalFloatLiteral(boolean isDotStart) {
+        int nextChar = peek();
+
+        // Direct process to the dot switch case
+        if (isDotStart) {
+            nextChar = LexerTerminals.DOT;
+        }
```

Comment:
can we get rid of this `isDotStart` flag? Why does this method needs to know whether the numeric started with a dot?

Reply:
What if this method always accepts the starting char as the input? Will that solve this complexity?

Reply:
the reason that I needed a `isDotStart` flag is that when `processDecimalFloatLiteral()`  is calling from `processDot()`, reader is already advanced for the `DOT`. But when the  `processDecimalFloatLiteral()` is calling from `processNumericalLiteral()`, reader is not advanced for the `DOT`. If we can fix either places we don't need a flag at all. Can't advance reader for `DOT` inside `processNumericalLiteral()` then it will cause problem for `ParserMode.IMPORT`.  Passing `starting char` won't help as what we want to know is not the starting char. we want want to know whether starting char is already advanced or not in the reader. 

As I said previously calling retreat() method inside processDot() would solve the problem. Then the `processDecimalFloatLiteral()` doesn't need to know whether it is calling from.

Any suggestion?